overlord/servicestate/servicestate.go

```diff
@@ -41,38 +42,141 @@ type Instruction struct {
 
 type ServiceActionConflictError struct{ error }
 
+// changeIDForContext returns change ID for non-ephemeral context
+// or empty string if context is nil or ephemeral.
+func changeIDForContext(context *hookstate.Context) string {
+	if context != nil && !context.IsEphemeral() {
+		if task, ok := context.Task(); ok {
+			if chg := task.Change(); chg != nil {
+				return chg.ID()
+			}
+		}
+	}
+	return ""
+}
+
+// serviceControlTs creates "service-control" task for every snap derived from appInfos.
+func serviceControlTs(st *state.State, appInfos []*snap.AppInfo, inst *Instruction) (*state.TaskSet, error) {
+	servicesBySnap := make(map[string][]string)
+	for _, app := range appInfos {
+		if !app.IsService() {
+			// this function should be called with services only
+			return nil, fmt.Errorf("internal error: %s is not a service", app.Name)
+		}
+		snapName := app.Snap.InstanceName()
+		servicesBySnap[snapName] = append(servicesBySnap[snapName], app.Name)
+	}
+
+	ts := state.NewTaskSet()
+	var prev *state.Task
+	for snapName, services := range servicesBySnap {
+		var snapst snapstate.SnapState
+		if err := snapstate.Get(st, snapName, &snapst); err != nil {
+			return nil, err
+		}
+		info, err := snapst.CurrentInfo()
+		if err != nil {
+			return nil, err
+		}
+		if len(info.Services()) == 0 {
+			return nil, fmt.Errorf("snap %q does not have services", snapName)
+		}
+
+		cmd := &ServiceAction{SnapName: snapName}
+		switch {
+		case inst.Action == "start":
+			cmd.Action = "start"
+			if inst.Enable {
+				cmd.ActionModifier = "enable"
+			}
+		case inst.Action == "stop":
+			cmd.Action = "stop"
+			if inst.Disable {
+				cmd.ActionModifier = "disable"
+			}
+		case inst.Action == "restart":
+			if inst.Reload {
+				cmd.Action = "reload-or-restart"
+			} else {
+				cmd.Action = "restart"
+			}
+		default:
+			return nil, fmt.Errorf("unknown action %q", inst.Action)
+		}
+
+		var svcs []string
+		for _, svcName := range services {
+			if svcName == snapName {
+				// all services of the snap
+				svcs = nil
+				break
+			}
+			app, ok := info.Apps[svcName]
+			if !(ok && app.IsService()) {
```

Comment:
Aren't we already checking `app.IsService()` in the loop above? In addition, the two conditions are not equivalent and could lead to a confusing error message (unknown application vs known non-service application).